File: KeeperLegacy/Models/Economy.swift
import Foundation

// MARK: - Player Economy State

struct PlayerEconomy: Codable {
    var coins: Int
    var stardust: Int

    // Lifetime stats for analytics / achievement tracking
    var totalCoinsEarned: Int
    var totalCoinsSpent: Int
    var totalStardustSpent: Int

    init(startingCoins: Int = 500, startingStardust: Int = 0) {
        self.coins              = startingCoins
        self.stardust           = startingStardust
        self.totalCoinsEarned   = startingCoins
        self.totalCoinsSpent    = 0
        self.totalStardustSpent = 0
    }

    // MARK: Transactions

    /// Returns true if the player can afford the given coin cost.
    func canAffordCoins(_ amount: Int) -> Bool {
        return coins >= amount
    }

    func canAffordStardust(_ amount: Int) -> Bool {
        return stardust >= amount
    }

    /// Spend coins. Returns false if insufficient funds.
    @discardableResult
    mutating func spendCoins(_ amount: Int) -> Bool {
        guard canAffordCoins(amount) else { return false }
        coins         -= amount
        totalCoinsSpent += amount
        return true
    }

    @discardableResult
    mutating func spendStardust(_ amount: Int) -> Bool {
        guard canAffordStardust(amount) else { return false }
        stardust          -= amount
        totalStardustSpent += amount
        return true
    }

    mutating func earnCoins(_ amount: Int) {
        coins            += amount
        totalCoinsEarned += amount
    }

    mutating func earnStardust(_ amount: Int) {
        stardust += amount
    }
}

// MARK: - Pricing Tables

/// Sourced from economy_system_v1.json
struct PricingTable {

    // MARK: Food
    struct Food {
        let id: String
        let name: String
        let coinCost: Int
        let hungerRestored: Double  // 0.0–1.0

        static let catalog: [Food] = [
            Food(id: "basic_kibble",   name: "Basic Kibble",   coinCost: 10, hungerRestored: 0.25),
            Food(id: "hearty_stew",    name: "Hearty Stew",    coinCost: 20, hungerRestored: 0.50),
            Food(id: "magical_feast",  name: "Magical Feast",  coinCost: 30, hungerRestored: 1.00),
        ]
    }

    // MARK: Breeding
    struct Breeding {
        /// Cost to breed two creatures of matching rarity
        static func cost(rarity: Rarity) -> Int {
            switch rarity {
            case .common:   return 300
            case .uncommon: return 800
            case .rare:     return 2500
            }
        }
    }

    // MARK: Customer Orders
    struct CustomerOrder {
        let id: UUID
        let requiredCreatureCatalogID: String
        let requiredRarity: Rarity
        let minHappiness: Double    // 0.0–1.0 threshold
        let coinReward: Int
        let expiresAt: Date
        var isFulfilled: Bool

        init(
            creatureCatalogID: String,
            rarity: Rarity,
            minHappiness: Double,
            coinReward: Int,
            expiresInHours: Double = 8
        ) {
            self.id                      = UUID()
            self.requiredCreatureCatalogID = creatureCatalogID
            self.requiredRarity          = rarity
            self.minHappiness            = minHappiness
            self.coinReward              = coinReward
            self.expiresAt               = Date().addingTimeInterval(expiresInHours * 3600)
            self.isFulfilled             = false
        }

        var isExpired: Bool { Date() >= expiresAt }

        /// Base reward range by rarity. Actual reward set at order creation.
        static func rewardRange(rarity: Rarity) -> ClosedRange<Int> {
            switch rarity {
            case .common:   return 50...200
            case .uncommon: return 200...500
            case .rare:     return 500...1500
            }
        }
    }

    // MARK: Sell Values
    /// Calculate how many coins a player gets for selling a creature.
    static func sellValue(rarity: Rarity, happinessMultiplier: Double) -> Int {
        let base = rarity.baseSellValue
        let multiplier = max(1.0, min(2.0, happinessMultiplier))
        return Int(Double(base) * multiplier)
    }
}

// MARK: - IAP / Stardust Packages

struct StardustPackage: Identifiable {
    let id: String
    let stardust: Int
    let usdPrice: Double
    let displayPrice: String    // e.g. "$4.99"

    static let packages: [StardustPackage] = [
        StardustPackage(id: "stardust_500",  stardust: 500,  usdPrice: 4.99,  displayPrice: "$4.99"),
        StardustPackage(id: "stardust_1000", stardust: 1000, usdPrice: 9.99,  displayPrice: "$9.99"),
        StardustPackage(id: "stardust_2500", stardust: 2500, usdPrice: 19.99, displayPrice: "$19.99"),
    ]
}

File: KeeperLegacyGodot/Models/Economy.cs
// Models/Economy.cs
// Port of Economy.swift
// Pure C# — no Godot dependency.

using System;
using System.Collections.Generic;

namespace KeeperLegacy.Models
{
    // ── Player Economy State ──────────────────────────────────────────────────

    public class PlayerEconomy
    {
        public int Coins { get; set; }
        public int Stardust { get; set; }

        // Lifetime stats
        public int TotalCoinsEarned { get; set; }
        public int TotalCoinsSpent { get; set; }
        public int TotalStardustSpent { get; set; }

        public PlayerEconomy(int startingCoins = 500, int startingStardust = 0)
        {
            Coins              = startingCoins;
            Stardust           = startingStardust;
            TotalCoinsEarned   = startingCoins;
            TotalCoinsSpent    = 0;
            TotalStardustSpent = 0;
        }

        // ── Transactions ──────────────────────────────────────────────────────

        public bool CanAffordCoins(int amount)    => Coins    >= amount;
        public bool CanAffordStardust(int amount) => Stardust >= amount;

        /// Spend coins. Returns false if insufficient funds.
        public bool SpendCoins(int amount)
        {
            if (!CanAffordCoins(amount)) return false;
            Coins          -= amount;
            TotalCoinsSpent += amount;
            return true;
        }

        /// Spend stardust. Returns false if insufficient funds.
        public bool SpendStardust(int amount)
        {
            if (!CanAffordStardust(amount)) return false;
            Stardust           -= amount;
            TotalStardustSpent += amount;
            return true;
        }

        public void EarnCoins(int amount)
        {
            Coins            += amount;
            TotalCoinsEarned += amount;
        }

        public void EarnStardust(int amount) => Stardust += amount;
    }

    // ── Pricing Tables ────────────────────────────────────────────────────────

    public static class PricingTable
    {
        // MARK: Food

        public class Food
        {
            public string Id { get; }
            public string Name { get; }
            public int CoinCost { get; }
            public double HungerRestored { get; }   // 0.0–1.0

            public Food(string id, string name, int coinCost, double hungerRestored)
            {
                Id             = id;
                Name           = name;
                CoinCost       = coinCost;
                HungerRestored = hungerRestored;
            }

            public static readonly List<Food> Catalog = new()
            {
                new Food("basic_kibble",  "Basic Kibble",  10, 0.25),
                new Food("hearty_stew",   "Hearty Stew",   20, 0.50),
                new Food("magical_feast", "Magical Feast", 30, 1.00),
            };
        }

        // MARK: Breeding

        public static class Breeding
        {
            /// Coin cost to breed two creatures of matching rarity.
            public static int Cost(Rarity rarity) => rarity switch
            {
                Rarity.Common   => 300,
                Rarity.Uncommon => 800,
                Rarity.Rare     => 2500,
                _               => 300
            };
        }

        // MARK: Customer Orders

        public class CustomerOrder
        {
            public Guid Id { get; }
            public string RequiredCreatureCatalogId { get; }
            public Rarity RequiredRarity { get; }
            public double MinHappiness { get; }     // 0.0–1.0 threshold
            public int CoinReward { get; }
            public DateTime ExpiresAt { get; }
            public bool IsFulfilled { get; set; }

            public CustomerOrder(
                string creatureCatalogId,
                Rarity rarity,
                double minHappiness,
                int coinReward,
                double expiresInHours = 8.0)
            {
                Id                       = Guid.NewGuid();
                RequiredCreatureCatalogId = creatureCatalogId;
                RequiredRarity           = rarity;
                MinHappiness             = minHappiness;
                CoinReward               = coinReward;
                ExpiresAt                = DateTime.UtcNow.AddHours(expiresInHours);
                IsFulfilled              = false;
            }

            // For deserialization
            public CustomerOrder(Guid id, string creatureCatalogId, Rarity rarity,
                double minHappiness, int coinReward, DateTime expiresAt, bool isFulfilled)
            {
                Id                        = id;
                RequiredCreatureCatalogId = creatureCatalogId;
                RequiredRarity            = rarity;
                MinHappiness              = minHappiness;
                CoinReward                = coinReward;
                ExpiresAt                 = expiresAt;
                IsFulfilled               = isFulfilled;
            }

            public bool IsExpired => DateTime.UtcNow >= ExpiresAt;

            /// Base reward range by rarity. Actual reward is set at order creation.
            public static (int min, int max) RewardRange(Rarity rarity) => rarity switch
            {
                Rarity.Common   => (50,  200),
                Rarity.Uncommon => (200, 500),
                Rarity.Rare     => (500, 1500),
                _               => (50,  200)
            };
        }

        // MARK: Sell Values

        /// Calculate coins earned for selling a creature.
        public static int SellValue(Rarity rarity, double happinessMultiplier)
        {
            int baseValue     = rarity.BaseSellValue();
            double multiplier = Math.Max(1.0, Math.Min(2.0, happinessMultiplier));
            return (int)(baseValue * multiplier);
        }
    }

    // ── IAP / Stardust Packages ───────────────────────────────────────────────

    public class StardustPackage
    {
        public string Id { get; }
        public int StardustAmount { get; }
        public double UsdPrice { get; }
        public string DisplayPrice { get; }

        public StardustPackage(string id, int stardustAmount, double usdPrice, string displayPrice)
        {
            Id             = id;
            StardustAmount = stardustAmount;
            UsdPrice       = usdPrice;
            DisplayPrice   = displayPrice;
        }

        public static readonly List<StardustPackage> Packages = new()
        {
            new StardustPackage("stardust_500",  500,  4.99,  "$4.99"),
            new StardustPackage("stardust_1000", 1000, 9.99,  "$9.99"),
            new StardustPackage("stardust_2500", 2500, 19.99, "$19.99"),
        };
    }
}
